Add Interceptors

simon-berger opened the pull request:
This is a draft to add Interceptos to CPR with minimal changes to the structure, as proposed in #719.
I would appreciate feedback on the proposed changes and suggestions for improvement, if any, before I fully develop the concept @COM8.
Interceptors in OkHttp: [link](https://square.github.io/okhttp/features/interceptors/)

**Interceptor class:**
Abstract base class which custom interceptors need to implement / inherit from. 
```c++
class Interceptor {
  public:
    virtual ~Interceptor() = default;

    // The intercept method gets automaticall called for each added interceptor and should
    // procceed execution with session.procceed()
    virtual Response intercept(Session& session) = 0;
};
```

**Sample Implementation of a custom Interceptor:**
The following custom interceptor would redirect any request to *"<server_base_url>/hello.html"* and rewrite the response url to the original requested one.
```c++
class HiddenHelloWorldRedirectInterceptor : public Interceptor {
  public:
    Response intercept(Session& session) override {
        // Save original url
        Url old_url = session.GetFullRequestUrl();

        // Rewrite the url
        Url url{server->GetBaseUrl() + "/hello.html"};
        session.SetUrl(url);

        // Procceed the chain
        Response response = session.proceed();

        // Restore the url again
        response.url = old_url;
        return response;
    };
};
```

**How to add a custom Interceptor:**
Adding a new interceptor can be done via session objects:
```c++
session.AddInterceptor(std::make_shared<HiddenHelloWorldRedirectInterceptor>());
```

**Changes in structure / Access specifier:**
Session would receive two new functions:
```c++
void AddInterceptor(const std::shared_ptr<Interceptor>& pinterceptor);
Response proceed();
```

Besides the implementation of the new session functions, the Session::Impl class now has a private queue of added interceptors and a reference to the respective Session object:
```c++
std::queue<std::shared_ptr<Interceptor>> interceptors;
Session* psession_;
```
This cyclic reference is currently necessary, as we need to pass a Session object to the intercept() method of potential added Interceptors.

Moreover it was necessary to make the following two functions of Session::Impl public, as they need to be called by Session:
```c++
Response makeRequest();
void prepareCommon();
```

**Some thoughts:**
- Instead of extending Session with proceed(), it would also be possible to implement a Interceptor::Chain class like it is done in OkHttp [link](https://github.com/square/okhttp/blob/3ad1912f783e108b3d0ad2c4a5b1b89b827e4db9/okhttp/src/jvmMain/kotlin/okhttp3/Interceptor.kt#L78)
  * However, this would require a little bit more changes (especially regarding acces specifiers)
- Interceptors might not change the request method (Get, Post, ...)
- The pointer handling should perhaps be improved

COM8 commented:
The general idea looks promising, but I won't have too much time until Thursday to give you a more detailed Feedback.

COM8 reviewed (changes requested):
In my eyes the general structure you are proposing looks good.
@sornian is this what you were thinking about when creating the feature request?

To keep it simple I would not implement something like `Interceptor::Chain`, if there is no specific need for it.

Regarding changing the request type I don't think this is easily possible with the current structure. One could spin up a custom session inside the interceptor call and since both classes are declared as friend, he would be able to change private stuff there in theory.

simon-berger commented:
@COM8 thank you for the feedback. I will take a closer look at it in the next few days / next week and work on it.

simon-berger commented on cpr/session.cpp:
```diff
@@ -825,6 +831,13 @@ void Session::Impl::prepareCommon() {
 
 Response Session::Impl::makeRequest() {
+    if (!interceptors.empty()) {
+        // At least one interceptor exists -> Ececute its intercept function
```
We actually call the makeRequest function multiple times (for each interceptor).

**Example:**
We add two interceptors to the queue:
  1. ChangeRequestUrlInterceptor: Interceptor which changes the request URL
  2. ChangeStatusCodeInterceptor: Interceptor which changes the response status code

Now when we make a Get request, we enter Session::Impl::makeRequest() where we pop and execute the first interceptor (ChangeRequestUrlInterceptor) from the queue. The intercept function of the interceptor modifies the request url and calls Session::proceed(), which again calls Session::Impl::makeRequest().

In Session::Impl::makeRequest() we pop and execute the first interceptor (ChangeStatusCodeInterceptor) again from the queue.

This process is repeated and the next time we enter Session::Impl::makeRequest(), the interceptor queue is empty and the (modified) request is executed.

The reponse object is then passed through the interceptors and modified by them if specified.

COM8 replied:
Ah, got it. Thanks!

simon-berger commented:
I have implemented the review and expanded the concept as follows:
1. I have created a "Friend" relationship between Session::Impl and Session, and between Session and Interceptor. This way I could make the Session::proceed(), Session::makeRequest() and Session::prepareCommon() functions private again.
2. Since the "friend" relationship does not transfer to inherited classes, the Interceptor class was extended by the function proceed(Session), which can now be called in the Intercept function instead of session.procced().
3. The Interceptor::proceed(Session, ProceedHttpMethod) function can now be used to change the http method of the request during interception.
4. I have added many tests

However, the msvc compiler builds are now failing and I don't really know why. @COM8 do you have any idea why these fail? My guess is that it might be the template function or the enum in the intercept. Unfortunately, I don't have any experience with msvc and had problems setting up the compiler locally.
Here is the relevant build-output:
```
...
session.cpp
  D:\a\cpr\cpr\include\cpr/interceptor.h(15,9): error C2143: syntax error: missing '}' before '(' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(15,9): error C2059: syntax error: 'constant' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(21,5): error C2143: syntax error: missing ';' before '}' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(21,5): error C2238: unexpected token(s) preceding ';' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,25): error C2588: '::~Interceptor': illegal global destructor [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,14): error C2575: 'Interceptor': only member functions and bases can be virtual [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,28): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,14): error C2610: 'int cpr::Interceptor(void)': is not a special member function or comparison operator which can be defaulted [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(24,22): error C2575: 'intercept': only member functions and bases can be virtual [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(24,52): error C2059: syntax error: 'constant' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(29,47): error C2061: syntax error: identifier 'ProceedHttpMethod' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(32,47): error C2061: syntax error: identifier 'ProceedHttpMethod' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(35,1): error C2059: syntax error: '}' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(35,1): error C2143: syntax error: missing ';' before '}' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\cpr\session.cpp(19,15): error C2143: syntax error: missing ';' before '{' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\cpr\session.cpp(19,15): error C2447: '{': missing function header (old-style formal list?) [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
    threadpool.cpp
    timeout.cpp
    unix_socket.cpp
    util.cpp
    response.cpp
    redirect.cpp
    Generating Code...
    Compiling...
    interceptor.cpp
  D:\a\cpr\cpr\include\cpr/interceptor.h(15,9): error C2143: syntax error: missing '}' before '(' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(15,9): error C2059: syntax error: 'constant' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(21,5): error C2143: syntax error: missing ';' before '}' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(21,5): error C2238: unexpected token(s) preceding ';' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,25): error C2588: '::~Interceptor': illegal global destructor [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,14): error C2575: 'Interceptor': only member functions and bases can be virtual [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,28): error C4430: missing type specifier - int assumed. Note: C++ does not support default-int [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(23,14): error C2610: 'int cpr::Interceptor(void)': is not a special member function or comparison operator which can be defaulted [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(24,22): error C2575: 'intercept': only member functions and bases can be virtual [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(24,52): error C2059: syntax error: 'constant' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(29,47): error C2061: syntax error: identifier 'ProceedHttpMethod' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(32,47): error C2061: syntax error: identifier 'ProceedHttpMethod' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(35,1): error C2059: syntax error: '}' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\include\cpr/interceptor.h(35,1): error C2143: syntax error: missing ';' before '}' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\cpr\interceptor.cpp(5,15): error C2143: syntax error: missing ';' before '{' [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
  D:\a\cpr\cpr\cpr\interceptor.cpp(5,15): error C2447: '{': missing function header (old-style formal list?) [D:\a\cpr\cpr\build\cpr\cpr.vcxproj]
    Generating Code...
  Error: The process 'C:\Program Files\CMake\bin\cmake.exe' failed with exit code 1
```

COM8 commented:
Grr, how I love msvc...
I can have look at it later today or tomorrow, when I have access to my Windows machine again. But I'd also say this probably has something to do with templates. msvc is not that complete in regards to templates.

simon-berger commented:
Ok, I will try to make some changes regarding the templates now and hopefully I can fix it by myself. Otherwise it would be great if you could have a look at it later 👍

COM8 commented:
Well, OK. This is an interesting one :D

simon-berger commented:
Docs: https://github.com/libcpr/docs/pull/18

COM8 reviewed (approved):
Works and looks good now. Thank you!
